File: web/app/src/types/order.ts
import { z } from "zod";
import { Datetime } from "./primitives/datetime";
import { Fraction } from "./primitives/fraction";
import { Uuid } from "./primitives/uuid";
import { Direction } from "./primitives/direction";

export const PrivateOrder = z.object({
  id: Uuid,
  created_at: Datetime,
  is_active: z.boolean(),
  quote_asset_id: Uuid,
  base_asset_id: Uuid,
  price: Fraction,
  quote_asset_volume: Fraction,
  quote_asset_volume_left: Fraction,
  maker_fee: Fraction,
  direction: Direction,
});
export type PrivateOrder = z.infer<typeof PrivateOrder>;

File: server/worker/src/database/projections/order.rs
use chrono::{DateTime, Utc};
use fraction::Fraction;
use serde::{Deserialize, Serialize};
use uuid::Uuid;

#[derive(Debug, Serialize, Deserialize, Clone)]
pub struct Order {
    pub id: Uuid,
    pub created_at: DateTime<Utc>,
    pub last_modification_at: DateTime<Utc>,
    pub user_id: Uuid,
    pub is_active: bool,
    pub quote_asset_id: Uuid,
    pub base_asset_id: Uuid,
    pub price: Fraction,
    pub quote_asset_volume: Fraction,
    pub quote_asset_volume_left: Fraction,
    pub maker_fee: Fraction,
}
